compiler/jsgen.nim

```diff
@@ -1855,7 +1855,10 @@ proc genArrayConstr(p: PProc, n: PNode, r: var TCompRes) =
   for i in countup(0, sonsLen(n) - 1):
     if i > 0: add(r.res, ", ")
     gen(p, n.sons[i], a)
-    add(r.res, a.res)
+    if a.typ == etyBaseIndex:
+      add(r.res, n.sons[i].sym.loc.r)
```

Comment:
Yeah, this is a total hack. I tested some more and it doesn't deal properly with copying at all. I must admit I'm quite confused about how ptrs/refs are dealt with. The second element seems to be used for indexing, but it's always just 0. Wouldn't it be better to simply use a single element array?

Looks to me like a bunch of stuff will have to be changed since even seemingly simple stuff seems to generate rubbish, eg. (on 0.19)
```nim
block:
    var f: ref string
    new(f)
    f[] = "abc"
    echo f[]
```
generates
```js
L5: do {
	var f_25079 = [[null, 0]];
	var f_25079_Idx = 0;
	f_25079[1] = null;
	f_25079[0][f_25079[1]] = nimCopy(null, makeNimstrLit("abc"), NTI138);
	rawEcho(f_25079[0][f_25079[1]]);
} while(false);
```
`genNew` default initializes f[1] instead of f[0][0], and then the string is stored under f[0]["null"]. Yes, the "null" string key. WTF?

Anyway, I'll look into this more tommorow.

Reply:
> The second element seems to be used for indexing, but it's always just 0.

Not really. In order to support the `ptr`/`ref` semantics a pointer is described by a base and a displacement since JS has no concept of pointers. A small example will surely help here, consider the following snippet:

```nim
type
  Foo = object
    field: string

var
  oj = Foo(field: "bar")
  x = addr oj.field
  aj = [1,2,3]
  y = addr aj[1]

echo x[]
echo y[]
```

This is compiled down to:

```js
var oj_25006 = [{field: makeNimstrLit("bar")}];
var x_25008 = [oj_25006[0], "field"];
var aj_25012 = [nimCopy(null, [1, 2, 3], NTI25009)];
var y_25014 = [aj_25012[0], 1];
rawEcho(x_25008[0][x_25008[1]]);
rawEcho(cstrToNimstr((y_25014[0][y_25014[1]])+""));
```

As you can see the second element keeps track of "where" we are inside of the object the first element points to.